themes/default/content/blog/pulumi-is-imperative-declarative-imperative/index.md

```diff
@@ -0,0 +1,221 @@
+Let me use the Pulumi architecture to highlight why I mention *imperative* twice.
+
+## Pulumi Architecture
+
+Here is the diagram from our [How Pulumi Works](/docs/intro/concepts/how-pulumi-works/) page:
+
+![Pulumi Architecture](/images/docs/reference/engine-block-diagram.png)
+
+You code your infrastructure in your preferred programming language. When you are done coding, you run `pulumi up` and the Pulumi CLI starts
+the language host for your selected programming language, as well as the required providers. The interaction between these 3 parts of
+the architecture results in the actual creation or modification of your infrastructure.
+
+### Language Host
+
+Under the hood, the Pulumi CLI does a lot of things, but one of the first actions is starting the language runtime which is configured in
+the `Pulumi.yaml` project file. Here is a small Python example:
+
+```python
+from pulumi_aws import s3
+
+bucket = s3.Bucket('bucket')
+
+for i in range(10):
+    s3.BucketObject(
+        f'object-{i}',
+        s3.BucketObjectArgs(
+            bucket=bucket.id,
+            key=str(i),
+        )
+    )
+```
+
+A Pulumi program models the to-be state of your infrastructure. If you read the program above, you can see that
+we define 11 resources as our to-be infrastructure:
+
+* 1 [AWS S3](https://www.pulumi.com/registry/packages/aws/api-docs/s3/bucket/) bucket
+* 10 [Objects](https://www.pulumi.com/registry/packages/aws/api-docs/s3/bucketobject/) in the bucket created in the previous step
+
+The stack itself is also modelled as a resource and it is the parent to all other resources.
+
+```sh
+$ pulumi up
+
+Updating (<masked>/dev)
+
+View Live: https://app.pulumi.com/<masked>/blog_code/dev/updates/1
+
+     Type                    Name           Status
+ +   pulumi:pulumi:Stack     blog_code-dev  created
+ +   ├─ aws:s3:Bucket        bucket         created
+ +   ├─ aws:s3:BucketObject  object-0       created
+ +   ├─ aws:s3:BucketObject  object-1       created
+ +   ├─ aws:s3:BucketObject  object-3       created
+ +   ├─ aws:s3:BucketObject  object-2       created
+ +   ├─ aws:s3:BucketObject  object-6       created
+ +   ├─ aws:s3:BucketObject  object-4       created
+ +   ├─ aws:s3:BucketObject  object-5       created
+ +   ├─ aws:s3:BucketObject  object-7       created
+ +   ├─ aws:s3:BucketObject  object-9       created
+ +   └─ aws:s3:BucketObject  object-8       created
+
+Resources:
+    + 12 created
+
+Duration: 13s
+```
+
+While this is definitely an imperative program, there is one important thing to understand: instantiating an
+`s3.Bucket`, `s3.BucketObject` or any other Pulumi resource should not be interpreted as an imperative creation
+of the resource in the language host. Behind the scenes, any resource instantiation in the language host triggers a
+`Register Resource` request to the Pulumi engine. All these `Register Resource` requests together form the
```

Comment:
I'd use the same wording we use in the How Pulumi Works page specifically in case someone starts exploring further.

```suggestion
`Register Resource` request to the Pulumi deployment engine. All these `Register Resource` requests together form the
```